File: 146-lru-cache/lru-cache.py
class Node:
    def __init__(self, val, key):
        self.key = key
        self.val= val
        self.prev=None
        self.next=None


class LRUCache:

    def __init__(self, capacity: int):
        self.cap = capacity
        self.cache_map ={}
        self.head =Node(0,0)
        self.tail =Node(0,0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove(self, node):
        prev, next = node.prev, node.next
        prev.next = next
        next.prev = prev
    def _insert(self,newnode):
        prev = self.tail.prev
        prev.next = newnode
        newnode.prev = prev
        self.tail.prev = newnode
        newnode.next = self.tail
        

    def get(self, key: int) -> int:
        if key in self.cache_map:
            node = self.cache_map[key]
            self._remove(node)
            newnode = Node(node.val, node.key)
            self._insert(newnode)
            self.cache_map[key]=newnode
            return newnode.val
        else:
            return -1
        
        

    def put(self, key: int, value: int) -> None:
        if key in self.cache_map:
            node = self.cache_map[key]
            self._remove(node)
            newnode = Node(value, key)
            self._insert(newnode)
            self.cache_map[key]=newnode
        else:
            newnode = Node(value,key)
            self._insert(newnode)
            self.cache_map[key]=newnode
            
        if len(self.cache_map) > self.cap:
            last_used = self.head.next
            self._remove(last_used)
            del self.cache_map[last_used.key]
            


        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146-lru-cache/lru-cache.go
type Node struct {
    key int
    val int
    next *Node
    prev *Node
}

type LRUCache struct {
    head *Node
    tail *Node
    cap int
    nodeMap map[int]*Node
    
}


func Constructor(capacity int) LRUCache {
    lru := LRUCache {
        head: &Node{key:-1,val: -1},
        tail: &Node{key: -1,val: -1},
        cap: capacity,
        nodeMap: make(map[int]*Node),
    }
    lru.head.next = lru.tail
    lru.tail.prev= lru.head
    return lru
    
}


func (this *LRUCache) Get(key int) int {
    v, ok:= this.nodeMap[key]
    if ok {
        this.removeNode(v)
        this.add_to_head(v)
        return v.val
    }
    return -1
    
}
func (this *LRUCache) removeNode(node *Node) {
    node.prev.next, node.next.prev = node.next, node.prev
}

func (this *LRUCache) add_to_head(node *Node) {
    node.next = this.head.next
    node.prev = this.head
    this.head.next.prev = node
    this.head.next = node
}


func (this *LRUCache) Put(key int, value int)  {
    v, ok :=this.nodeMap[key]
    if ok  {
        v.val = value
        this.removeNode(v)
        this.add_to_head(v)

    } else 
    {
        if len(this.nodeMap) >= this.cap {
            lru:=this.tail.prev
            
        delete(this.nodeMap, lru.key)
        this.removeNode(lru)
        
        }
        node:= &Node{key:key,val: value}
        this.add_to_head(node) 
        this.nodeMap[key]=node

    }

    
}


/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */